feat: Add SpEL support to RetryWithTimeOutStrategy

Redouane-E opened the pull request:
- Update RetryWithTimeOutStrategy to evaluate SpEL expressions in timeOut and retryDelay parameters
- Add tests to verify SpEL expression evaluation and logging
- Ensure proper handling of mixed SpEL and non-SpEL values
- Improve logging to show evaluated SpEL expressions
![image](https://github.com/user-attachments/assets/37e70be7-4a35-4033-b6e3-dc78f6c6f47a)
![image](https://github.com/user-attachments/assets/86884ebd-844c-419e-adb8-65ce793432bc)

Commit:
feat: Add SpEL support to RetryWithTimeOutStrategy

KarimGl commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -106,6 +112,38 @@ public Status execute(ScenarioExecution scenarioExecution,
+        return value.contains("#{") || value.contains("${") || value.startsWith("#");
+    }
+
+    private Object evaluateExpression(String expression, StepDataEvaluator evaluator, ScenarioContext scenarioContext) {
```
- `String expression` -> `String timeout`
- No need to check if it's an spel (already done [here](https://github.com/chutney-testing/chutney/blob/main/chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/engine/evaluation/Strings.java#L56)). You can remove this method and call evaluate directly
` Object evaluatedTimeOut = evaluator.evaluate(...)`

Redouane-E replied:
For 
```
String expression -> String timeout
```
I named it "expression" to be generic, as it could represent either a timeout or a retry delay.

KarimGl commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -106,6 +112,38 @@ public Status execute(ScenarioExecution scenarioExecution,
+        step.addInformation("Retry strategy definition : [timeOut " + timeOutLog + "] [delay " + retryDelayLog + "]");
+    }
+
+    private boolean isSpelExpression(String value) {
```
can be remove.
See coment on `evaluateExpression `method

KarimGl commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -106,6 +112,38 @@ public Status execute(ScenarioExecution scenarioExecution,
+        return expression;
+    }
+
+    private long convertToMilliseconds(Object value) {
```
May be suppose that `value` param should respect `Duration` format (string so).
If agree with me, youcan remove this method and re use `toMilliSeconds`

KarimGl commented on chutney/engine/src/test/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategyTest.java:
```diff
@@ -317,4 +316,97 @@ private void stopExecution(ScenarioExecution scenarioExecution) {
+    }
+
+    @Test
+    public void should_handle_mixed_spel_and_non_spel_values() {
```
It can be removed because this is already checked in `should_evaluate_spel_expression_in_retryDelay `and `should_evaluate_spel_expression_in_timeout`

KarimGl commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -106,6 +112,38 @@ public Status execute(ScenarioExecution scenarioExecution,
         return st;
     }
 
+    private void logStrategyDefinition(Step step, String timeOut, String retryDelay, Object evaluatedTimeOut, Object evaluatedRetryDelay) {
```
You can remove this method. See coment on `evaluateExpression ` method

nbrouand commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -67,15 +69,20 @@ public Status execute(ScenarioExecution scenarioExecution,
-        Long timeOutMs = toMilliSeconds(timeOut);
-        Long retryDelayMs = toMilliSeconds(retryDelay);
-        Long timeLeft = timeOutMs;
+
```
remove empty line pls

nbrouand commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -67,15 +69,20 @@ public Status execute(ScenarioExecution scenarioExecution,
+        Object evaluatedTimeOut = evaluateExpression(timeOut, step.dataEvaluator(), scenarioContext);
+        Object evaluatedRetryDelay = evaluateExpression(retryDelay, step.dataEvaluator(), scenarioContext);
+
+        long timeOutMs = toMilliSeconds(evaluatedTimeOut.toString());
```
merge evaluateExpression and toMilliSeconds in one method named, by example : resolveDurationInMilliseconds

nbrouand commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -106,6 +113,19 @@ public Status execute(ScenarioExecution scenarioExecution,
+            Object result = evaluator.evaluate(expression, scenarioContext);
+            return result != null ? result : expression;
+        } catch (EvaluationException e) {
+            System.out.println("Failed to evaluate expression: " + expression);
```
No System.out pls

nbrouand commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -106,6 +113,19 @@ public Status execute(ScenarioExecution scenarioExecution,
         return st;
     }
 
+    private Object evaluateExpression(String expression, StepDataEvaluator evaluator, ScenarioContext scenarioContext) {
```
you dont need this method which is redondant with StepDataEvaluator. Just call evaluator.evaluate(expression, scenarioContext)

boddissattva replied:
Be careful here, the context to use to evaluate the expression should be the merge context from `scenarioContext` and `localContext` i think, as this strategy could be used into a for strategy loop.

boddissattva replied:
Maybe it could be cool that this strategy put a `retryCount` variable into the context that could be used.
To discuss on a new feature maybe.

nbrouand commented on chutney/engine/src/test/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategyTest.java:
```diff
@@ -317,4 +316,79 @@ private void stopExecution(ScenarioExecution scenarioExecution) {
     }
 
+    @Test
+    public void should_evaluate_spel_expression_in_timeout() {
```
Please do only one test with a @ParameterizedTest

boddissattva replied:
I would add that the assertion on info is out of the thing to test here i think. The feature has to test that the stepEvaluator of the step is used, and only that. Just let the test be in error if the `when` clause is not called. And if you **really** want to have an assertion, just verify this call then.

boddissattva commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -106,6 +109,28 @@ public Status execute(ScenarioExecution scenarioExecution,
+        try {
+            ScenarioContextImpl mergedContext = new ScenarioContextImpl();
+            mergedContext.putAll(scenarioContext);
+            mergedContext.putAll(localContext);
```
I would put this merge process out of this method and use directly a `Map<String, Object>` instead of a `ScenarioContextImpl`.
Besides, it will be executed once.

boddissattva commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -106,6 +109,28 @@ public Status execute(ScenarioExecution scenarioExecution,
     }
 
+    private long evaluateAndConvertToMilliseconds(String expression, StepDataEvaluator evaluator, ScenarioContext scenarioContext, Map<String, Object> localContext) {
+        if (evaluator == null) {
```
Is it a real use case ? I don't think that the step evaluator can be null.

boddissattva commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -106,6 +109,28 @@ public Status execute(ScenarioExecution scenarioExecution,
+            mergedContext.putAll(localContext);
+
+            Object result = evaluator.evaluate(expression, mergedContext);
+            String evaluatedExpression = result != null ? result.toString() : expression;
```
I don't think that the evaluation could be null.
In other classes, when we evaluate, we generally cast the result as the class we want. Here, we could cast the result as `String`.

boddissattva commented on chutney/engine/src/main/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategy.java:
```diff
@@ -106,6 +109,28 @@ public Status execute(ScenarioExecution scenarioExecution,
+            return toMilliSeconds(evaluatedExpression);
+        } catch (EvaluationException e) {
+            throw new RuntimeException("Failed to evaluate expression: " + expression, e);
+        }
```
We generally don't catch the `EvaluateException` which is already a runtime exception.
Let it be.

boddissattva commented on chutney/engine/src/test/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategyTest.java:
```diff
@@ -278,6 +280,36 @@ public void should_add_information_about_strategy_and_retries(Status[] stepStatu
 
+    @ParameterizedTest
+    @MethodSource("spelExpressionTestCases")
+    void should_evaluate_spel_expressions_in_strategy_properties(String timeOut, String retryDelay, String expectedTimeOut, String expectedRetryDelay) {
```
I'd like to see a test for the context and local context merge feature.

boddissattva commented on chutney/engine/src/test/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategyTest.java:
```diff
@@ -278,6 +280,36 @@ public void should_add_information_about_strategy_and_retries(Status[] stepStatu
+        Step step = mock(Step.class);
+        when(step.strategy()).thenReturn(Optional.of(strategyDefinition));
+
+        StepDataEvaluator evaluator = mock(StepDataEvaluator.class);
```
Why not use a real evaluator ? it's not so complicated to build.

boddissattva commented on chutney/engine/src/test/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategyTest.java:
```diff
@@ -278,6 +280,36 @@ public void should_add_information_about_strategy_and_retries(Status[] stepStatu
+
+        verify(evaluator).evaluate(eq(timeOut), any());
+        verify(evaluator).evaluate(eq(retryDelay), any());
+        verify(step, atLeastOnce()).addInformation(anyString());
```
redundant with the argument captor, no ?

boddissattva commented on chutney/engine/src/test/java/com/chutneytesting/engine/domain/execution/strategies/RetryWithTimeOutStrategyTest.java:
```diff
@@ -278,6 +280,36 @@ public void should_add_information_about_strategy_and_retries(Status[] stepStatu
+        strategyUnderTest.execute(ScenarioExecution.createScenarioExecution(null), step, new ScenarioContextImpl(), null);
+
+        verify(evaluator).evaluate(eq(timeOut), any());
+        verify(evaluator).evaluate(eq(retryDelay), any());
```
Mockito throws an exception i think if the when conditions are not met.

boddissattva commented:
Patch from mob review.
[patch-203.patch](https://github.com/user-attachments/files/17414168/patch-203.patch)

Redouane-E commented:
I've refactored the code to ensure that the UI properly displays the information message. Also, I've added an assertion in the parameterized test to validate it.
**Before:** 
![image](https://github.com/user-attachments/assets/f80b8913-6017-4946-bd16-204ae989a728)
**After:**
![image](https://github.com/user-attachments/assets/f6414803-08c7-4ae7-8362-af4b1ce0ee45)

Once approved, i'll squash the commit

boddissattva commented:
@Redouane-E 
Some minor changes if you want.
[chore()__Clean_code_+_Add_information_test_with_spEL_evaluation.patch](https://github.com/user-attachments/files/17434192/chore.__Clean_code_%2B_Add_information_test_with_spEL_evaluation.patch)
